### PR TITLE
Complex offer pages

### PR DESCRIPTION
**Pull request by smdalton**
# Complex Offer Pages
Three views were added to properly separate the different functionalities for a transaction which is one of three types.

## overview
1. All three views use the same html template, but feed it with different information depending on type of transaction.

2. The type of the transaction is encoded in the link for a specific item on the item_details page, where it is discovered from the DB and passed in for rendering. Clicking on the link will send a request to one of three routes, each of which will process a specific type of transaction

3. Each item details page will have a picture of the item being bid on/bought/free

Additionally, the template will render differently in the following ways

## Views
___

#### 1. Trade transactions:

* all snag button clicks on a **trade transaction item** link to the route:
transaction/trade_only/<**item_id**>, where item id is a valid id from a cash_only item in the DB
which directs them to the **view** trade_transaction

* will load a screen of the currently logged in user's inventory and allow them to select items via checkbox for trade offer submissions 

* All Trade transactions are handled by the view trade_transaction

* when the users submits the trade, any item that loaded from their inventory that has a checkbox which is clicked, will be an item that is used to create a single TradeOffer database row, all of these rows are saved to the database **on a successful submission, creating 0-n database entries where n is the number of selected items.**

* comment not strictly required, if present will be saved, if not no comment will be saved
___

#### 2. Cash Transactions:
* all snag button clicks on a **cash only item** link to the route:
transaction/cash_only/<**item_id**>, where item id is a valid id from a cash_only item in the DB
which directs them to the **view** cash_transaction

* The users will be supplied with a cash amount box, where they can place a cash offer on the item for snagging

* All trade transactions are handled by the view trade_transaction

* comment not strictly required, if present will be saved, if not no comment will be saved
___

#### 3. Free Transactions:
* all snag button clicks on an item of transaction type **free** link to the route:
transaction/free/<**item_id**>, where item id is a valid id from a cash_only item in the DB
which directs them to the **view** cash_transaction

* All free transactions are handled by the view free_transaction

* comment not strictly required, if present will be saved, if not no comment will be saved

#### Edge cases

1. In the event that a user attempts to access a route outside of the normal link clicking process, say by manually entering an item_id in the browser and performing a GET request:

2. The cash/trade/free routes should reject all attempts to view items that are not associated with that routes category, i.e. (trying to manually **access the trade_only route with a cash item**

3. if a user attempts to look up an id that is not in the database, they should be gracefully redirected to the home page

### CODE REVIEW

File: slug_trade/slug_trade/settings.py [hunk 1]
```diff
@@ -32,13 +32,15 @@
 # Application definition
 
 INSTALLED_APPS = [
+    # 'livereload',
```
Comment (smdalton) [inline slug_trade/slug_trade/settings.py]: package that causes browser refresh on saving of a file (very useful tool for front end dev, disabled for testing)

File: slug_trade/slug_trade/settings.py [hunk 2]
```diff
@@ -50,6 +52,7 @@
     'django.contrib.auth.middleware.SessionAuthenticationMiddleware',
     'django.contrib.messages.middleware.MessageMiddleware',
     'django.middleware.clickjacking.XFrameOptionsMiddleware',
+    # 'livereload.middleware.LiveReloadScript'
```
Comment (smdalton) [inline slug_trade/slug_trade/settings.py]: package that causes browser refresh on saving of a file (very useful tool for front end dev, disabled for testing)

File: slug_trade/slug_trade_app/admin.py [hunk 3]
```diff
@@ -25,7 +25,7 @@ def get_inline_instances(self, request, obj=None):
 class ItemAdmin(admin.ModelAdmin):
     search_fields = ['name']
     list_filter = ['category', 'trade_options']
-    list_display = ['name', 'get_name', 'price', 'category', 'trade_options', 'bid_counter', 'condition']
+    list_display = ['id', 'name', 'get_name', 'price', 'category', 'trade_options', 'bid_counter', 'condition']
```
Comment (smdalton) [inline slug_trade/slug_trade_app/admin.py]: Needed to alter this to display an item's id in the admin pane as well as it's trade type ignorer to select items to test the routes with

File: slug_trade/slug_trade_app/admin.py [hunk 4]
```diff
@@ -56,32 +56,32 @@ def get_name(self, obj):
 
 
 class OfferCommentAdmin(admin.ModelAdmin):
-    list_display = ['get_item_name']
```
Comment (smdalton) [inline slug_trade/slug_trade_app/admin.py]: caused crashes in Django admin pane so it was removed

File: slug_trade/slug_trade_app/forms.py [hunk 5]
```diff
@@ -21,14 +22,39 @@ class Meta():
         )
 
 
+class CashTransactionForm(forms.ModelForm):
```
Comment (smdalton) [inline slug_trade/slug_trade_app/forms.py]: Forms that were created for each of the views for transactions and comments

File: slug_trade/slug_trade_app/models.py [hunk 6]
```diff
@@ -34,10 +34,10 @@
 )
 
 TRADE_OPTIONS = (
-    ('0','Cash Only'),
```
Comment (smdalton) [inline slug_trade/slug_trade_app/models.py]: style enforcement changes

File: slug_trade/slug_trade_app/views.py [hunk 7]
```diff
@@ -20,24 +23,24 @@ def index(request):
     test = "This was passed from the backend!"
     if debug:
         print("in index view")
-    return render(request, 'slug_trade_app/index.html',{'test':test})
+    return render(request, 'slug_trade_app/index.html', {'test': test})
 
 
 def products(request):
     categories = [
```
Comment (smdalton) [inline slug_trade/slug_trade_app/views.py]: style enforcement whitespace changes

File: slug_trade/slug_trade_app/views.py [hunk 8]
```diff
@@ -290,10 +538,10 @@ def signup(request):
             created_profile = profile_form.save(commit=False)
 
             profile = UserProfile(
-                user = created_user,
-                profile_picture = request.FILES['profile_picture'],
-                bio = created_profile.bio,
-                on_off_campus = created_profile.on_off_campus
```
Comment (smdalton) [inline slug_trade/slug_trade_app/views.py]: whitespace style changes

**Comment by AustinKellar** [comment]
Sidney and I worked together to solve the merge conflicts. The PR should be good to start testing

**Comment by egomish** [comment]
There is no way to access the item details page via the app. Testing item details is unfeasible until the item details page can be reached via the app (rather than manipulating the URL).

**Comment by smdalton** [comment]
I will add a link from the products page so that the user can be directed within the app to the item_details page.

**Pull request merged**